File: cinema-tickets-java/src/main/java/uk/gov/dwp/uc/pairtest/domain/TicketTypeRequest.java
package uk.gov.dwp.uc.pairtest.domain;

/**
 * Immutable Object
 */

public class TicketTypeRequest {

    private final int noOfTickets;
    private final Type type;

    public TicketTypeRequest(Type type, int noOfTickets) {
        this.type = type;
        this.noOfTickets = noOfTickets;
    }

    public int getNoOfTickets() {
        return noOfTickets;
    }

    public Type getTicketType() {
        return type;
    }

    public enum Type {
        ADULT(20), CHILD(10) , INFANT(0);

        private final int cost;

        Type(int cost) {
            this.cost = cost;
        }

        public int getCost() {
            return cost;
        }
    }

}

File: cinema-tickets-javascript/src/pairtest/lib/TicketTypeRequest.js
/**
 * Immutable Object.
 */

import InvalidPurchaseException from "./InvalidPurchaseException";

export default class TicketTypeRequest {
  /** @type {String} String value of ticket type */
  #type;

  /** @type {Number} Number value representing number of tickets. */
  #noOfTickets;

  constructor(type, noOfTickets) {
    if (!this.#Type.includes(type)) {
      throw new TypeError(
        `type must be ${this.#Type
          .slice(0, -1)
          .join(", ")}, or ${this.#Type.slice(-1)}`
      );
    }

    if (!Number.isInteger(noOfTickets)) {
      throw new TypeError("noOfTickets must be an integer");
    }

    if (noOfTickets < 0) {
      throw new InvalidPurchaseException(
        "ticketNumberError",
        "Numbers of tickets purchased must be 0 or greater."
      );
    }

    this.#type = type;
    this.#noOfTickets = noOfTickets;
  }

  /**
     * Getter which returns number of tickets.
     * 
     * @return {Number} #noOfTickets - the number of tickets.
     */
  getNoOfTickets() {
    return this.#noOfTickets;
  }

  /**
     * Getter which returns type of ticket.
     * 
     * @return {String} #type - thetype of ticket.
     */
  getTicketType() {
    return this.#type;
  }

  /** @type {Array.<String>} Array containing strings representing ticket types */
  #Type = ["ADULT", "CHILD", "INFANT"];
}
